<a href="https://colab.research.google.com/github/Shritej24c/NLP101/blob/main/HW/HW8/ShritejShrikant_Chavan_HW8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Data Preparation

### Import Packages

In [1]:
# check if we are using google colab
from pathlib import Path
import textwrap
if 'google.colab' in str(get_ipython()):
    from google.colab import drive
    drive.mount("/content/drive")
    !pip install datasets transformers bertopic umap hdbscan  openai -U -qq

    base_folder = Path("/content/drive/MyDrive/NLP")
else:
    base_folder = Path("/home/harpreet/Insync/google_drive_shaannoor/data")

Mounted at /content/drive
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 521.2/521.2 kB 7.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.2/8.2 MB 39.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 154.1/154.1 kB 13.0 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.2/5.2 MB 39.6 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 221.4/221.4 kB 28.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 115.3/115.3 kB 16.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 20.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 90.9/90.9 kB 11.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.0/86.0 kB 10.8 MB/s eta 0:00:00
  Preparing metadata (s

In [2]:
def print_wrap(text, width=80):
    """
    Prints the given text, wrapping lines to a maximum of the specified width (default is 80 characters).

    Args:
    text (str): The text to be printed.
    width (int): The maximum width of a line, in characters.
    """
    wrapper = textwrap.TextWrapper(width=width)
    wrapped_text = wrapper.fill(text)
    print(wrapped_text)

In [3]:
data_folder = base_folder/'datasets/hw8'
model_folder = base_folder/'models/nlp_fall_2023/hw8'
model_folder.mkdir(exist_ok=True, parents=True)
data_folder.mkdir(exist_ok=True, parents=True)

### Load Dataset

In [4]:
from datasets import load_dataset

medium_dataset = load_dataset('csv', data_files= str(data_folder /'titles_cleaned.csv'))

Extracting data files:   0%|          | 0/1 [00:00<?, ?it/s]

Generating train split: 0 examples [00:00, ? examples/s]

In [5]:
medium_dataset

DatasetDict({
    train: Dataset({
        features: ['Unnamed: 0', 'url', 'title', 'claps', 'responses', 'readTime', '$oid', '$date', 'datetime', 'year', 'month', 'date', 'hour', 'minute', 'second'],
        num_rows: 28383
    })
})

In [6]:
medium_dataset['train'].features

{'Unnamed: 0': Value(dtype='int64', id=None),
 'url': Value(dtype='string', id=None),
 'title': Value(dtype='string', id=None),
 'claps': Value(dtype='float64', id=None),
 'responses': Value(dtype='float64', id=None),
 'readTime': Value(dtype='string', id=None),
 '$oid': Value(dtype='string', id=None),
 '$date': Value(dtype='string', id=None),
 'datetime': Value(dtype='string', id=None),
 'year': Value(dtype='float64', id=None),
 'month': Value(dtype='float64', id=None),
 'date': Value(dtype='float64', id=None),
 'hour': Value(dtype='float64', id=None),
 'minute': Value(dtype='float64', id=None),
 'second': Value(dtype='float64', id=None)}

In [7]:
medium_titles = medium_dataset['train']['title']

## Sentence BERT for embeddings

In [8]:
from sentence_transformers import SentenceTransformer

# We load our model
embedding_model = SentenceTransformer('all-mpnet-base-v2')

.gitattributes:   0%|          | 0.00/1.18k [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.6k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/571 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

data_config.json:   0%|          | 0.00/39.3k [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/438M [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/239 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/363 [00:00<?, ?B/s]

train_script.py:   0%|          | 0.00/13.1k [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

In [9]:
import joblib

In [10]:
#embeddings = embedding_model.encode(medium_titles)
#joblib.dump(embeddings, model_folder/'medium_nlp_abstract_embeddings')
embeddings_medium = joblib.load(model_folder/'medium_nlp_abstract_embeddings')

In [11]:
embeddings_medium.shape

(28383, 768)

## UMAP

In [12]:
from umap import UMAP
from hdbscan import HDBSCAN

In [13]:
umap_model = UMAP(n_neighbors=15, n_components=5, min_dist=0.0, metric='cosine', random_state=21)

##HDBSCAN

In [14]:
hdbscan_model = HDBSCAN(min_cluster_size=30, metric='euclidean', cluster_selection_method='eom', prediction_data=True)

## CountVectorizer

In [15]:
from sklearn.feature_extraction.text import CountVectorizer
vectorizer_model = CountVectorizer(stop_words="english", min_df=2, ngram_range=(1, 2))

In [67]:
# Creating the BERTopic model with embedding model

# topic_model = BERTopic(embedding_model=embedding_model, umap_model=umap_model, hdbscan_model=hdbscan_model, vectorizer_model=vectorizer_model)
# topics, probs = topic_model.fit_transform(documents=abstracts)
from bertopic import BERTopic
# Creating the BERTopic model without embedding model
topic_model = BERTopic(embedding_model=embedding_model, umap_model=umap_model,
                       hdbscan_model=hdbscan_model, vectorizer_model=vectorizer_model,
                       calculate_probabilities=True)
topics, probs = topic_model.fit_transform(embeddings=embeddings_medium, documents=medium_titles)


### Interpreting Topics

#### Overall Summary of Topics

In [68]:
len(topics)

28383

In [94]:
topic_model.get_topic_info().head(20)

Topic  Count                                               Name  \
0      -1  10064                        -1_job_design_earn_business   
1       0    712               0_designs_design_designing_designers   
2       1    599          1_newsletters_emails_newsletter_marketing   
3       2    447  2_startups_startup_entrepreneurs_entrepreneurship   
4       3    396                          3_sales_b2b_marketing_b2c   
5       4    375                4_data_analyst_analytics_analytical   
6       5    367              5_branding_brands_rebranding_rebrands   
7       6    362                          6_song_jessica_ryan_chris   
8       7    359                 7_linkedin_jobs_recruiters_company   
9       8    300                  8_adwords_ads_adsense_advertising   
10      9    269            9_marketing_advertising_social_marketer   
11     10    262                     10_women_men_dating_patriarchy   
12     11    258            11_midjourney_guide_style_illustrations   
13     12    257                    12_seo_marketing_ranking_google   
14     13    249            13_products_management_product_managers   
15     14    246  14_segmentation_thresholding_opencv_convolutional   
16     15    245       15_tensorflow_detector_segmentations_pytorch   
17     16    244              16_marketing_advertising_marketer_seo   
18     17    240         17_copywriting_copywriters_copywriter_copy   
19     18    226   18_diffusion_stablediffusion_denoising_diffusers   

                                       Representation  \
0   [job, design, earn, business, build, model, ai...   
1   [designs, design, designing, designers, ux, tr...   
2   [newsletters, emails, newsletter, marketing, m...   
3   [startups, startup, entrepreneurs, entrepreneu...   
4   [sales, b2b, marketing, b2c, business, ecommer...   
5   [data, analyst, analytics, analytical, study, ...   
6   [branding, brands, rebranding, rebrands, brand...   
7   [song, jessica, ryan, chris, music, who, wildf...   
8   [linkedin, jobs, recruiters, company, recruite...   
9   [adwords, ads, adsense, advertising, ad, campa...   
10  [marketing, advertising, social, marketer, bus...   
11  [women, men, dating, patriarchy, masculinity, ...   
12  [midjourney, guide, style, illustrations, imag...   
13  [seo, marketing, ranking, google, wordpress, g...   
14  [products, management, product, managers, mana...   
15  [segmentation, thresholding, opencv, convoluti...   
16  [tensorflow, detector, segmentations, pytorch,...   
17  [marketing, advertising, marketer, seo, digita...   
18  [copywriting, copywriters, copywriter, copy, w...   
19  [diffusion, stablediffusion, denoising, diffus...   

                                  Representative_Docs  
0   [Make Money Online Listening to Music - Evan C...  
1   [Make UX simple again - UX Collective, What th...  
2   [The Best 3 Email Scraping Tools That Help Ema...  
3   [Is Your Startup a Good Fit for Venture Capita...  
4   [B2B sales needs a major overhaul - Sandhya Do...  
5   [Becoming a “real” data analyst - Towards Data...  
6   [The Brand Archetype Strategy - Umair Asghar -...  
7   [If You Try To Do Everything, You Won’t Do Any...  
8   [5 Tools for LinkedIn Marketing - Sangita Cham...  
9   [Digital Marketing (Google Ads Definition — Co...  
10  [Developing a Social Media Marketing Strategy ...  
11  [Meet the Female Dating Strategy: Women Teachi...  
12  [The Ultimate Guide to MidJourney V5.1 (with i...  
13  [SEO Guide: To Show You Path for SEO - Chandni...  
14  [The Five Whys of a Product Manager - Kubo - M...  
15  [Introduction to Image Processing with Python,...  
16  [DEtection TRansformer (DETR) vs. YOLO for obj...  
17  [Digital Marketing - Moose - Medium, What is D...  
18  [on copywriting, on copywriting, What About Co...  
19  [Stable Diffusion as an API, Stable Diffusion,...

In [70]:
dd = topic_model.get_topic_info()

In [71]:
dd.shape

(169, 5)

In [72]:
dd[dd['Topic'] == 126 ]

Topic  Count                          Name  \
127    126     48  126_father_car_daughter_left   

                                        Representation  \
127  [father, car, daughter, left, mother, child, m...   

                                   Representative_Docs  
127  [The Grace Period for Being a Recent Widow Has...

In [73]:
topic_model.get_topic(0)

[('ux', 0.06250497792800819),
 ('design', 0.05944338737515278),
 ('collective', 0.04268497037656447),
 ('ux collective', 0.04232014009392014),
 ('ui', 0.028005796164668837),
 ('ux design', 0.027299277203722248),
 ('designer', 0.025242210069824682),
 ('designers', 0.023231444709548417),
 ('ux planet', 0.01865767003122706),
 ('planet', 0.01830805794747912)]

In [74]:
topic_model.get_topic(124)

[('android', 0.350888540130032),
 ('android apps', 0.07982195451190098),
 ('android interview', 0.053005984616721624),
 ('activity', 0.050965894186584926),
 ('architecture', 0.04843630404000443),
 ('architect', 0.03982334840092273),
 ('interview questions', 0.03974533546622055),
 ('context', 0.037912810350334655),
 ('testing', 0.037887147521482346),
 ('app architecture', 0.037092110716688276)]

In [75]:
topic_model.visualize_distribution(probs[0])

In [76]:
medium_titles[1]

'Generative search experience: the next step in search engine UX?'

#### Approximate Topic Distribution

For more details, refer to https://maartengr.github.io/BERTopic/getting_started/distribution/distribution.html


In [77]:
topic_distr, _ = topic_model.approximate_distribution(medium_titles)

In [78]:
topic_model.visualize_distribution(topic_distr[1])

#### Interactive Visualizations

In [79]:
# Calculate the topic distributions on a token-level
topic_distr, topic_token_distr = topic_model.approximate_distribution(medium_titles, calculate_tokens=True)

# Visualize the token-level distributions
df = topic_model.visualize_approximate_distribution(medium_titles[1], topic_token_distr[1])
df


In [80]:
topics[1]

-1

The above document should have been clustered in the generative AI topic rather than search engine. It's appalling that the word generative hasn't been given weightage to model this document into appropriate topic of GEN AI

In [81]:
topic_model.get_document_info(medium_titles)

Document  Topic  \
0      5 ChatGPT Prompts I Use to Write High-Ranking ...     -1   
1      Generative search experience: the next step in...     -1   
2      What Google’s Generative AI Announcement Means...     -1   
3      Google’s Generative AI Updates Will Shatter Or...     -1   
4      Will Generative AI Kill SEO? Here’s What an SE...     -1   
...                                                  ...    ...   
28378  A Successful Freelance Copywriter Addresses th...     61   
28379  6 Engaging YouTube Thumbnail Design Ideas That...     -1   
28380                          What’s New in Next.js 14?     49   
28381  ChatGPT Has Changed My Approach to Learning Ne...     25   
28382  Using ChatGPT For Web Scraping: A Practical Guide    116   

                                               Name  \
0                       -1_medium_chatgpt_data_life   
1                       -1_medium_chatgpt_data_life   
2                       -1_medium_chatgpt_data_life   
3                       -1_medium_chatgpt_data_life   
4                       -1_medium_chatgpt_data_life   
...                                             ...   
28378          61_ai_ai writing_generators_ai write   
28379                   -1_medium_chatgpt_data_life   
28380                         49_js_js 13_13_nextjs   
28381  25_chatgpt_medium chatgpt_using chatgpt_chat   
28382          116_scraping_web scraping_web_scrape   

                                          Representation  \
0      [medium, chatgpt, data, life, using, ai, make,...   
1      [medium, chatgpt, data, life, using, ai, make,...   
2      [medium, chatgpt, data, life, using, ai, make,...   
3      [medium, chatgpt, data, life, using, ai, make,...   
4      [medium, chatgpt, data, life, using, ai, make,...   
...                                                  ...   
28378  [ai, ai writing, generators, ai write, writing...   
28379  [medium, chatgpt, data, life, using, ai, make,...   
28380  [js, js 13, 13, nextjs, app, dynamic, material...   
28381  [chatgpt, medium chatgpt, using chatgpt, chat,...   
28382  [scraping, web scraping, web, scrape, medium w...   

                                     Representative_Docs  \
0      [Make Money Online Listening to Music - Evan C...   
1      [Make Money Online Listening to Music - Evan C...   
2      [Make Money Online Listening to Music - Evan C...   
3      [Make Money Online Listening to Music - Evan C...   
4      [Make Money Online Listening to Music - Evan C...   
...                                                  ...   
28378  [7 Best Free AI Tools for Content Creation and...   
28379  [Make Money Online Listening to Music - Evan C...   
28380  [Understanding the SSR, SSG, and ISR in Next.j...   
28381  [ChatGPT 與資料視覺化 - twdsmeetup - Medium, What’s ...   
28382  [Web Scraping and Web Automation with Python 2...   

                                             Top_n_words  Probability  \
0      medium - chatgpt - data - life - using - ai - ...     0.517531   
1      medium - chatgpt - data - life - using - ai - ...     0.445157   
2      medium - chatgpt - data - life - using - ai - ...     0.392761   
3      medium - chatgpt - data - life - using - ai - ...     0.478484   
4      medium - chatgpt - data - life - using - ai - ...     0.392761   
...                                                  ...          ...   
28378  ai - ai writing - generators - ai write - writ...     0.698075   
28379  medium - chatgpt - data - life - using - ai - ...     0.956274   
28380  js - js 13 - 13 - nextjs - app - dynamic - mat...     0.456850   
28381  chatgpt - medium chatgpt - using chatgpt - cha...     0.245311   
28382  scraping - web scraping - web - scrape - mediu...     0.275113   

       Representative_document  
0                        False  
1                        False  
2                        False  
3                        False  
4                        False  
...                        ...  
28378                    False  

In [82]:
topic_model.visualize_topics()

In [83]:
topic_model.visualize_barchart(top_n_topics=20)

In [84]:
topic_model.visualize_barchart(topics=list(range(50, 58, 1)))

### Hierarchical Topics
In BERTopic for creating a hierarchical structure of topics, the default choice is Scipy's ward linkage function. Nonetheless, depending on your specific requirements, you might prefer other linkage functions like single, complete, average, centroid, or median. BERTopic offers the flexibility to customize the linkage function and even select the desired distance function to suit your particular use case.


In [85]:
from scipy.cluster import hierarchy as sch

# Hierarchical topics
# linkage_function = lambda x: sch.linkage(x, 'single', optimal_ordering=True)
# hierarchical_topics = topic_model.hierarchical_topics(docs, linkage_function=linkage_function)

hierarchical_topics = topic_model.hierarchical_topics(medium_titles)
topic_model.visualize_hierarchy(hierarchical_topics=hierarchical_topics)


100%|██████████| 167/167 [00:01<00:00, 127.61it/s]


Following are the major topics into which the documents are divided:

1. UI/UX
2. ChatGPT
3. Generative AI
4. Google Ads
5. social Media following
6. Email Marketing
7. Big Data

These are just some few examples of clusters which are divided further into subclusters


In [86]:
topic_model.topic_representations_

{-1: [('medium', 0.005275896206273433),
  ('chatgpt', 0.00518205698583761),
  ('data', 0.004786524368214274),
  ('life', 0.004722417685825071),
  ('using', 0.004604612433825728),
  ('ai', 0.004364726776236341),
  ('make', 0.004253299534447538),
  ('google', 0.0040877865016399),
  ('science', 0.0038601033299912177),
  ('data science', 0.00364780686384404)],
 0: [('ux', 0.06250497792800819),
  ('design', 0.05944338737515278),
  ('collective', 0.04268497037656447),
  ('ux collective', 0.04232014009392014),
  ('ui', 0.028005796164668837),
  ('ux design', 0.027299277203722248),
  ('designer', 0.025242210069824682),
  ('designers', 0.023231444709548417),
  ('ux planet', 0.01865767003122706),
  ('planet', 0.01830805794747912)],
 1: [('email', 0.10096506389619882),
  ('email marketing', 0.06558259092129835),
  ('newsletter', 0.057532295180489156),
  ('emails', 0.03268372462660052),
  ('substack', 0.03162205228756058),
  ('email list', 0.030878460859256326),
  ('list', 0.028821951404442398),
  

In [87]:
# Save original representations
from copy import deepcopy
original_topics = deepcopy(topic_model.topic_representations_)

In [88]:
def compare_topic_changes(new_model, original_topic_words, max_length=75, top_n_topics=10):

    """Displays differences in top words of topic representations between the original and new models."""

    for topic_id in range(top_n_topics):
        # Extract top 5 words per topic from the original and new models
        original_top_words = "_".join(word for word, _ in original_topic_words[topic_id][:5])
        new_top_words = "_".join(word for word, _ in new_model.get_topic(topic_id)[:5])

        # Calculate whitespace for alignment
        alignment_spaces = " " * (max_length - len(original_top_words))

        # Print 'before' and 'after' topic word changes
        print(f"Topic: {topic_id}    {original_top_words}{alignment_spaces} >>     {new_top_words}")


## KeyBERTInspired




In [89]:
# KeyBERTInspired
from bertopic.representation import KeyBERTInspired
representation_model = KeyBERTInspired()

# Update our topic representations
topic_model.update_topics(medium_titles, representation_model=representation_model)


### Comparison of Topics

In [90]:
# Show topic differences
compare_topic_changes(topic_model, original_topics)

Topic: 0    ux_design_collective_ux collective_ui                                       >>     designs_design_designing_designers_ux
Topic: 1    email_email marketing_newsletter_emails_substack                            >>     newsletters_emails_newsletter_marketing_mail
Topic: 2    startup_founders_venture_entrepreneur_startups                              >>     startups_startup_entrepreneurs_entrepreneurship_founders
Topic: 3    b2b_sales_marketing_b2b sales_funnel                                        >>     sales_b2b_marketing_b2c_business
Topic: 4    data_data engineering_data analyst_data science_analyst                     >>     data_analyst_analytics_analytical_study
Topic: 5    brand_branding_identity_brand identity_logo                                 >>     branding_brands_rebranding_rebrands_brand
Topic: 6    medium_jessica_holiday medium_ryan holiday_holiday                          >>     song_jessica_ryan_chris_music
Topic: 7    linkedin_profile_leads_recruiters_usin

In [91]:
topic_model.get_topic(0, full=True)

{'Main': [('designs', 0.53380334),
  ('design', 0.51747537),
  ('designing', 0.50500005),
  ('designers', 0.48914835),
  ('ux', 0.47295764),
  ('trends', 0.43599612),
  ('innovation', 0.39386082),
  ('industry', 0.36847156),
  ('designer', 0.36549464),
  ('future', 0.36333477)]}

In [92]:
# topic_model.topic_representations_

### Hierarchical Topics

In [93]:
hierarchical_topics = topic_model.hierarchical_topics(medium_titles)
topic_model.visualize_hierarchy(hierarchical_topics=hierarchical_topics)

100%|██████████| 167/167 [00:28<00:00,  5.95it/s]


Above you can see 167 meaningful clusters/topics, with around 10k documents as outliers. In the following code, I am attempting to remove outliers based on the probabilities strategy

### Remove Outliers

In [95]:
new_topics = topic_model.reduce_outliers(medium_titles, topics, probabilities=probs, strategy="probabilities")

In [96]:
# prompt: get values counts from elements of a list

from collections import Counter
c = Counter(new_topics)


In [97]:
print(c)
print(len(c))

Counter({0: 972, 2: 695, 3: 648, 6: 616, 1: 599, 4: 510, 125: 489, 5: 435, 126: 418, 12: 417, 10: 405, 8: 388, 7: 383, 39: 337, 26: 329, 11: 327, 25: 323, 19: 306, 9: 302, 14: 302, 21: 300, 16: 296, 13: 293, 110: 287, 15: 273, 83: 270, 119: 258, 35: 250, 20: 249, 38: 247, 17: 244, 68: 242, 18: 234, 32: 233, 99: 232, 111: 229, 28: 226, 31: 221, 162: 216, 146: 209, 24: 208, 41: 203, 50: 202, 33: 200, 42: 199, 30: 194, 144: 187, 46: 187, 65: 186, 34: 184, 22: 184, 55: 182, 64: 181, 23: 180, 40: 172, 95: 171, 27: 170, 36: 169, 82: 168, 29: 168, 37: 165, 120: 162, 44: 162, 72: 161, 59: 160, 43: 156, 56: 155, 54: 155, 87: 154, 58: 151, 98: 150, 62: 144, 135: 144, 57: 141, 88: 141, 47: 141, 48: 138, 49: 138, 104: 138, 149: 136, 115: 135, 84: 133, 61: 131, 101: 131, 60: 131, 143: 131, 51: 129, 45: 128, 86: 125, 105: 123, 70: 123, 52: 123, 92: 121, 93: 121, 165: 120, 157: 119, 91: 116, 74: 114, 53: 113, 77: 109, 166: 109, 141: 109, 89: 107, 148: 106, 63: 105, 113: 105, 80: 104, 158: 102, 151: 1

In [100]:
topic_model.update_topics(medium_titles, topics=new_topics)

2023-12-13 23:44:28,976 - BERTopic - WARNING: Using a custom list of topic assignments may lead to errors if topic reduction techniques are used afterwards. Make sure that manually assigning topics is the last step in the pipeline.Note that topic embeddings will also be created through weightedc-TF-IDF embeddings instead of centroid embeddings.


In [101]:
topic_model.get_topic_info().head(20)

Topic  Count                                      Name  \
0       0    972                 0_ux_design_collective_ui   
1       1    599        1_email_newsletter_emails_substack   
2       2    695  2_startup_entrepreneur_founders_startups   
3       3    648           3_b2b_sales_marketing_customers   
4       4    510        4_data_science_engineering_analyst   
5       5    435            5_brand_branding_identity_logo   
6       6    616               6_medium_the_jessica_poetry   
7       7    383             7_linkedin_410_deleted_author   
8       8    388                   8_ads_google_adwords_ad   
9       9    302       9_social_media_marketing_management   
10     10    405           10_women_men_woman_relationship   
11     11    327           11_midjourney_prompts_images_v5   
12     12    417            12_seo_website_traffic_ranking   
13     13    293    13_product_manager_management_managers   
14     14    302           14_image_opencv_computer_vision   
15     15    273           15_object_detection_yolov8_yolo   
16     16    296       16_digital_marketing_agency_company   
17     17    244   17_copywriting_copywriter_copy_handbook   
18     18    234       18_diffusion_stable_controlnet_sdxl   
19     19    306                      19_ai_art_dall_image   

                                       Representation  \
0   [ux, design, collective, ui, designers, design...   
1   [email, newsletter, emails, substack, list, su...   
2   [startup, entrepreneur, founders, startups, in...   
3   [b2b, sales, marketing, customers, funnel, cus...   
4   [data, science, engineering, analyst, engineer...   
5   [brand, branding, identity, logo, brands, coff...   
6   [medium, the, jessica, poetry, thief, taylor, ...   
7   [linkedin, 410, deleted, author, profile, lead...   
8   [ads, google, adwords, ad, campaign, advertisi...   
9   [social, media, marketing, management, strateg...   
10  [women, men, woman, relationship, dating, tran...   
11  [midjourney, prompts, images, v5, create, cons...   
12  [seo, website, traffic, ranking, search, rank,...   
13  [product, manager, management, managers, coali...   
14  [image, opencv, computer, vision, python, pyto...   
15  [object, detection, yolov8, yolo, yolov7, mode...   
16  [digital, marketing, agency, company, presence...   
17  [copywriting, copywriter, copy, handbook, watk...   
18  [diffusion, stable, controlnet, sdxl, stabledi...   
19  [ai, art, dall, image, images, mlearning, gene...   

                                  Representative_Docs  
0   [Make UX simple again - UX Collective, What th...  
1   [The Best 3 Email Scraping Tools That Help Ema...  
2   [Is Your Startup a Good Fit for Venture Capita...  
3   [B2B sales needs a major overhaul - Sandhya Do...  
4   [Becoming a “real” data analyst - Towards Data...  
5   [The Brand Archetype Strategy - Umair Asghar -...  
6   [If You Try To Do Everything, You Won’t Do Any...  
7   [5 Tools for LinkedIn Marketing - Sangita Cham...  
8   [Digital Marketing (Google Ads Definition — Co...  
9   [Developing a Social Media Marketing Strategy ...  
10  [Meet the Female Dating Strategy: Women Teachi...  
11  [The Ultimate Guide to MidJourney V5.1 (with i...  
12  [SEO Guide: To Show You Path for SEO - Chandni...  
13  [The Five Whys of a Product Manager - Kubo - M...  
14  [Introduction to Image Processing with Python,...  
15  [DEtection TRansformer (DETR) vs. YOLO for obj...  
16  [Digital Marketing - Moose - Medium, What is D...  
17  [on copywriting, on copywriting, What About Co...  
18  [Stable Diffusion as an API, Stable Diffusion,...  
19  [5 AI Art Generators You Should Know About - M...

The outliers have been reduced to 972

In [102]:
hierarchical_topics = topic_model.hierarchical_topics(medium_titles)
topic_model.visualize_hierarchy(hierarchical_topics=hierarchical_topics)

100%|██████████| 167/167 [00:01<00:00, 130.17it/s]


### Merge Topics

In [103]:
topics_to_merge = [[105, 25, 95, 120 ], [44, 62, 163], [124, 160], [142, 127]]

topic_model.merge_topics(medium_titles, topics_to_merge)

In the 1st list merged topics similar to ChatGPT, 2nd list related to podcasts, 3rd list about Android and 4th list about note taking

In [104]:
hierarchical_topics = topic_model.hierarchical_topics(medium_titles)
topic_model.visualize_hierarchy(hierarchical_topics=hierarchical_topics)

100%|██████████| 160/160 [00:01<00:00, 121.96it/s]
